PlayerScripts/PlayerScriptsPlace/StarterPlayer/StarterPlayerScripts/ControlScript/MasterControl.rbxmx

```diff
@@ -12,95 +12,115 @@
-	if character then
-		if CachedHumanoid and CachedHumanoid.Parent == character then
+	if LocalCharacter then
+		if CachedHumanoid then
 			return CachedHumanoid
 		else
-			CachedHumanoid = nil
-			for _,child in pairs(character:GetChildren()) do
-				if child:IsA('Humanoid') then
-					CachedHumanoid = child
-					return CachedHumanoid
-				end
-			end
+			CachedHumanoid = LocalCharacter:FindFirstChildOfClass("Humanoid")
+			return CachedHumanoid
 		end
 	end
 end
 
+VRService.Changed:connect(function(prop)
+	if prop == "VREnabled" then
+		VREnabled = VRService.VREnabled
+	end
+end)
+
+local characterAncestryChangedConn = nil
+local characterChildRemovedConn = nil
+local function characterAdded(character)
+	if characterAncestryChangedConn then
+		characterAncestryChangedConn:disconnect()
+	end
+
+	if characterChildRemovedConn then
+		characterChildRemovedConn:disconnect()
+	end	
+	
+	LocalCharacter = character
+	CachedHumanoid = LocalCharacter:FindFirstChildOfClass("Humanoid")
+	characterAncestryChangedConn = character.AncestryChanged:connect(function()
+		if not character:IsDescendantOf(workspace) then
+			LocalCharacter = nil
+		else
+			LocalCharacter = character
+		end
+	end)
+	
+	characterChildRemovedConn = character.ChildRemoved:connect(function(child)
+		if child == CachedHumanoid then
+			CachedHumanoid = nil
+		end
+	end)
+end
+
+if LocalCharacter then
+	characterAdded(LocalCharacter)
+end
+LocalPlayer.CharacterAdded:connect(characterAdded)
+
 --[[ Public API ]]--
 function MasterControl:Init()
-	
+
+	local getHumanoid = self.GetHumanoid
+	local moveFunc = LocalPlayer.Move
 	local renderStepFunc = function()
-		if LocalPlayer and LocalPlayer.Character then
-			local humanoid = self:GetHumanoid()
-			if not humanoid then return end
-			
-			if humanoid and not humanoid.PlatformStand and isJumping then
-				local state = humanoid:GetState()
-				if state ~= Enum.HumanoidStateType.Jumping and state ~= Enum.HumanoidStateType.Freefall and state ~= Enum.HumanoidStateType.Landed then
-					humanoid.Jump = isJumping
-				end
-			end
+		local humanoid = getHumanoid()
+		if not humanoid then return end
 			
-			local adjustedMoveValue = moveValue
-			if HasVRAPI and UserInputService.VREnabled and workspace.CurrentCamera.HeadLocked then
-				local vrFrame = UserInputService.UserHeadCFrame
-				local lookVector = Vector3.new(vrFrame.lookVector.X, 0, vrFrame.lookVector.Z).unit
-				local rotation = CFrame.new(Vector3.new(0, 0, 0), lookVector)
-				adjustedMoveValue = rotation:vectorToWorldSpace(adjustedMoveValue)
+		if isJumping and not humanoid.PlatformStand then
+			local state = humanoid:GetState()
+			if state ~= STATE_JUMPING and state ~= STATE_FREEFALL and state ~= STATE_LANDED then
+				humanoid.Jump = isJumping
 			end
-			
-			moveFunc(LocalPlayer, adjustedMoveValue, true)	
-			
-			
 		end
+		
+		local adjustedMoveValue = moveValue
+		if VREnabled and workspace.CurrentCamera.HeadLocked then
+			local vrFrame = VRService:GetUserCFrame()
```

Comment:
UserHeadCFrame is deprecated in favor of this method now, I'll cache the lookup though